File: java/bank-account/src/main/java/BankAccount.java
public class BankAccount {
	int balance;
	boolean enabled;

	void open() {
		enabled = true;
	}

	int getBalance() throws BankAccountActionInvalidException {
		checkEnabled();

		return balance;
	}

	synchronized void deposit(int amount) throws BankAccountActionInvalidException {
		checkEnabled();
		checkAmount(amount);

		balance += amount;
	}

	synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
		checkEnabled();
		checkAmount(amount);

		if (balance == 0) {
			throw new BankAccountActionInvalidException("Cannot withdraw money from an empty account");
		}
		if (amount > balance) {
			throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");
		}

		balance -= amount;
	}

	private void checkAmount(int amount) throws BankAccountActionInvalidException {
		if (amount < 0) {
			throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
		}
	}

	private void checkEnabled() throws BankAccountActionInvalidException {
		if (!enabled) {
			throw new BankAccountActionInvalidException("Account closed");
		}
	}

	void close() {
		enabled = false;
	}
}

File: python/bank-account/bank_account.py
import threading


class BankAccount(object):
    def __init__(self):
        self.lock = threading.Lock()
        self.balance = 0
        self.enabled = False

    def get_balance(self):
        self.check_enabled()

        return self.balance

    def open(self):
        if self.enabled:
            raise ValueError('Account is already opened!')

        self.enabled = True

    def deposit(self, amount):
        with self.lock:
            self.check_enabled()
            self.check_amount(amount)

            self.balance += amount

    def withdraw(self, amount):
        with self.lock:
            self.check_enabled()
            self.check_amount(amount)

            if amount > self.balance:
                raise ValueError('Amount can not exceed balance!')

            self.balance -= amount

    def close(self):
        if not self.enabled:
            raise ValueError('Account is already closed!')

        self.balance = 0
        self.enabled = False

    def check_enabled(self):
        if not self.enabled:
            raise ValueError('Account is closed!')

    def check_amount(self, amount):
        if amount < 0:
            raise ValueError('Amount can not be negative!')
